### PR TITLE
Auto detect GitPath

### PR DESCRIPTION
**Pull request by Grynn**
Hi,

The attached code auto-detects the GitPath, which allows making the GitPath attribute of the SemVerGitHarvester optional instead of required. This could simplify usage.

Auto-detection is basically just guessing the path - I check the system %PATH% and some well known locations like %ProgramFiles%\git\bin etc.

### CODE REVIEW

File: SemVerHarvester/SemVerGitHarvester.cs [hunk 1]
```diff
@@ -41,8 +44,27 @@ public SemVerGitHarvester(GitDescribeRunner gitDescribeRunner)
         /// <summary>
         ///     Gets or sets the path to the Git executable.
         /// </summary>
-        [Required]
-        public string GitPath { get; set; }
+        public string GitPath 
+        { 
+            get 
+            {
+#if NET40
```
Comment (jennings) [inline SemVerHarvester/SemVerGitHarvester.cs]: Can we use String.IsNullOrEmpty(this.gitPath.Trim()) instead so we don't need the preprocessor directives, and get the same behavior on different frameworks?

File: SemVerHarvester/SemVerGitHarvester.cs [hunk 2]
```diff
@@ -134,5 +156,52 @@ private void SetPropertiesFromGitDescribe(string gitDescribeResult)
...
+
+        [SuppressMessage("Microsoft.StyleCop.CSharp.SpacingRules", "SA1001:CommasMustBeSpacedCorrectly", Justification = "Formatting works better with #ifdef")]
+        private string FindGitExe()
+        {
+            var checkDirs = new List<string>();
+            checkDirs.AddRange(Environment.GetEnvironmentVariable("PATH").Split(';'));
+            checkDirs.AddRange(new[] 
+            {
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), @"git\bin")
+#if NET40
```
Comment (jennings) [inline SemVerHarvester/SemVerGitHarvester.cs]: I think it's undesireable to have different behavior on different frameworks.. can we do something that doesn't involve a preprocessor directive?

File: SemVerHarvester/SemVerGitHarvester.cs [hunk 3]
```diff
@@ -134,5 +156,52 @@ private void SetPropertiesFromGitDescribe(string gitDescribeResult)
...
+            p.StartInfo.CreateNoWindow = true;
+            p.Start();
+
+            string output = p.StandardOutput.ReadToEnd();
+            
+            p.WaitForExit();
+            return output;
+        }
+
+        [SuppressMessage("Microsoft.StyleCop.CSharp.SpacingRules", "SA1001:CommasMustBeSpacedCorrectly", Justification = "Formatting works better with #ifdef")]
```
Comment (jennings) [inline SemVerHarvester/SemVerGitHarvester.cs]: If we remove the preprocessor directive (see comment on line 187), we can remove this exception.

File: SemVerHarvester.Test/SemVerGitHarvesterTests.cs [hunk 4]
```diff
@@ -25,9 +25,10 @@ public class SemVerGitHarvesterTests
 
         /// <summary>
         ///     Confirms that Execute fails if GitPath is not set.
+        ///     Test fails if git.exe is not in path
         /// </summary>
         [Test]
-        public void Execute_returns_false_if_git_path_is_not_set()
+        public void Auto_detect_git_path()
```
Comment (jennings) [inline SemVerHarvester.Test/SemVerGitHarvesterTests.cs:31]: I think this actually needs to be split into separate tests for different scenarios:
- Test where Git is found in %ProgramFiles%\Git\bin\git.exe
- Test where Git is found in %ProgramFiles(x86)%\Git\bin\git.exe
- Test where Git is not found in either location

As it is, this no longer verifies that Execute() fails if Git.exe isn't found.

**Comment by jennings** [comment]
Wow, I didn't expect to actually get contributions, that's fantastic.

I remember wanting to auto-detect the Git location, so this is great. I do have a couple concerns, though. It looks like this change will behave differently on different frameworks, since we're calling IsNullOrWhitespace on .NET 4.0 and IsNullOrEmpty on other frameworks. I'm also concerned that we no longer test that a missing git.exe leads to Execute() returning false.

Do you have any thoughts? Perhaps I've misunderstood the change.

File: SemVerHarvester/SemVerGitHarvester.cs (hunk 1 above)
Comment (Grynn) [inline SemVerHarvester/SemVerGitHarvester.cs]: Sure; though I have some thoughts -

1) Conceptually: We'd have to backport features from .net 40 to .net 20 or live with the "least common denominator" in terms of functionality. I'm not sure if that's the best way to proceed, but hey ... you're the repo owner :). Also I've tried to be very careful and ensure that we don't have very different functionality between different frameworks, but we do take advantage of newer features, where available.

2) The pre-processor #defines are being auto-generated in the .csproj... so there's not much chance of someone defining NET40 while building for .net 20.... 

3) Nitpick: We'd need something other than 'this.gitPath.Trim()' (if gitPath is null then the Trim() will throw)... 

Lemme know if you want me to make a new commit - without the #if statements....

File: SemVerHarvester.Test/SemVerGitHarvesterTests.cs (hunk 4 above)
Comment (Grynn) [inline SemVerHarvester.Test/SemVerGitHarvesterTests.cs:31]: Agreed, I don't have much experience with unit tests, so this is great advice. I'll make a new commit ...

**Comment by Grynn** [comment]
What would be great is if we could check for the existence of 'IsNullOrWhitespace' so it would work with say .net 4.5 as well....

File: SemVerHarvester/SemVerGitHarvester.cs (hunk 1 above)
Comment (jennings) [inline SemVerHarvester/SemVerGitHarvester.cs]: It's just my own personal preferences, I don't really like preprocessor directives if they can be avoided.

And yeah, you're definitely right about the NullReferenceException, didn't really think it through. Maybe we can solve both problems with something like `String.IsNullOrEmpth( (gitPath ?? "").Trim() )` ? It's more cumbersome, but then the same code works for both frameworks.

**Comment by jennings** [comment]
@Grynn I just realized I never followed up with this. I don't think I really have the time to work on this anymore, so I made you a collaborator on this repo so you can changes if you want to.

**Comment by Grynn** [comment]
Kewl! Thanks, I have not looked at this in sometime, but will try and commit ...

**Pull request merged**